src/drivers/uavcan_v1/templates/_composite_type.j2

```diff
@@ -0,0 +1,115 @@
+{{ composite_type | definition_begin }}
+{
+{%- if T.fixed_port_id is not None %}
+	static constexpr CanardPortID PORT_ID = {{T.fixed_port_id}};
+{%- endif %}
+
+{%- set total_size = namespace(value=0) -%}
+{%- for field in composite_type.fields -%}
+    {% set total_size.value = total_size.value + field.data_type.bit_length_set|max %}
+{%- endfor %}
+	static constexpr size_t SIZE = {{total_size.value}};
+
+	static constexpr auto getDataTypeFullName() { return "{{T}}"; }
+
+{% for constant in composite_type.constants %}
+	static constexpr {{ constant.data_type | declaration }} {{ constant.name | id }} = {{ constant.value.native_value.numerator }} / {{ constant.value.native_value.denominator }};
+{%- endfor -%}
+
+{% if composite_type is UnionType %}
+#error "TODO: UnionType
+{% else %}
+
+{% for field in composite_type.fields -%}
+    {%- if field is not padding %}
+        {%- if field.data_type is FloatType %}
+            {%- if field.data_type.bit_length_set|max <= 32 %}
+	float {{ field | id }}{NAN};
+            {%- else %}
+	double {{ field | id }}{NAN};
+            {%- endif %}
+        {%- elif field.data_type is BooleanType %}
+	bool {{ field | id }}{false};
+        {%- else %}
+	{{ field.data_type | declaration }} {{ field | id }}{};
+        {%- endif -%}
+    {%- endif -%}
+{%- endfor %}
+{% endif %}
+
+	void serializeToBuffer(uint8_t* const buffer, const size_t starting_bit = 0)
```

Comment:
1. you need to provide the buffer size in bytes and check for overruns.
2. "size_t can store the maximum size [in bytes] of a theoretically possible object of any type (including array)." (from cppreference.com) Using it to store the total bit offset is theoretically incorrect since SIZE_T_MAX / 8 < LARGEST_POSSIBLE_ADDRESS_ON_SYSTEM. While this seems academic on 32 and 64-bit systems, on an 8-bit system that defined size_t as a uint16 (e.g. AVR8) this means you cannot have a buffer larger than ~8k. We could choose to use a 32-bit value since it's probably unreasonable to expect these routines to be working with 500MB buffers or larger but we'd be chewing through a lot more operations on 8-bit MCUs dealing with two instructions per read and write (again, assuming AVR8 which has 16-bit, atomic operations) of this field. I wonder if we should use two values, one a size_t starting_byte and the other a size_t starting_bit_in_byte to ensure there are no ambiguities nor capacity issues with this code?